LUONGTD/Open-vSwitch/project/vswitchd/ovs-vswitchd.conf.db.5
'\" p
.\" -*- nroff -*-
.TH "ovs-vswitchd.conf.db" 5 " DB Schema 7.15.1" "Open vSwitch 2.9.2" "Open vSwitch Manual"
.fp 5 L CR              \\" Make fixed-width font available as \\fL.
.de TQ
.  br
.  ns
.  TP "\\$1"
..
.de ST
.  PP
.  RS -0.15in
.  I "\\$1"
.  RE
..
.SH NAME
ovs-vswitchd\[char46]conf\[char46]db \- Open_vSwitch database schema
.PP
.PP
.PP
.PP
A database with this schema holds the configuration for one Open vSwitch daemon\[char46] The top-level configuration for the daemon is the \fBOpen_vSwitch\fR table, which must have exactly one record\[char46] Records in other tables are significant only when they can be reached directly or indirectly from the \fBOpen_vSwitch\fR table\[char46] Records that are not reachable from the \fBOpen_vSwitch\fR table are automatically deleted from the database, except for records in a few distinguished ``root set\(cq\(cq tables\[char46]
.SS "Common Columns"
.PP
.PP
Most tables contain two special columns, named \fBother_config\fR and \fBexternal_ids\fR\[char46] These columns have the same form and purpose each place that they appear, so we describe them here to save space later\[char46]
.RS
.TP
\fBother_config\fR: map of string-string pairs
Key-value pairs for configuring rarely used features\[char46] Supported keys, along with the forms taken by their values, are documented individually for each table\[char46]
.IP
A few tables do not have \fBother_config\fR columns because no key-value pairs have yet been defined for them\[char46]
.TP
\fBexternal_ids\fR: map of string-string pairs
Key-value pairs for use by external frameworks that integrate with Open vSwitch, rather than by Open vSwitch itself\[char46] System integrators should either use the Open vSwitch development mailing list to coordinate on common key-value definitions, or choose key names that are likely to be unique\[char46] In some cases, where key-value pairs have been defined that are likely to be widely useful, they are documented individually for each table\[char46]
.RE
.SH "TABLE SUMMARY"
.PP
The following list summarizes the purpose of each of the tables in the
\fBOpen_vSwitch\fR database.  Each table is described in more detail on a later
page.
.IP "Table" 1in
Purpose
.TQ 1in
\fBOpen_vSwitch\fR
Open vSwitch configuration\[char46]
.TQ 1in
\fBBridge\fR
Bridge configuration\[char46]
.TQ 1in
\fBPort\fR
Port configuration\[char46]
.TQ 1in
\fBInterface\fR
One physical network device in a Port\[char46]
.TQ 1in
\fBFlow_Table\fR
OpenFlow table configuration
.TQ 1in
\fBQoS\fR
Quality of Service configuration
.TQ 1in
\fBQueue\fR
QoS output queue\[char46]
.TQ 1in
\fBMirror\fR
Port mirroring\[char46]
.TQ 1in
\fBController\fR
OpenFlow controller configuration\[char46]
.TQ 1in
\fBManager\fR
OVSDB management connection\[char46]
.TQ 1in
\fBNetFlow\fR
NetFlow configuration\[char46]
.TQ 1in
\fBSSL\fR
SSL configuration\[char46]
.TQ 1in
\fBsFlow\fR
sFlow configuration\[char46]
.TQ 1in
\fBIPFIX\fR
IPFIX configuration\[char46]
.TQ 1in
\fBFlow_Sample_Collector_Set\fR
Flow_Sample_Collector_Set configuration\[char46]
.TQ 1in
\fBAutoAttach\fR
AutoAttach configuration\[char46]
.\" check if in troff mode (TTY)
.if t \{
.bp
.SH "TABLE RELATIONSHIPS"
.PP
The following diagram shows the relationship among tables in the
database.  Each node represents a table.  Tables that are part of the
``root set'' are shown with double borders.  Each edge leads from the
table that contains it and points to the table that its value
represents.  Edges are labeled with their column names, followed by a
constraint on the number of allowed values: \fB?\fR for zero or one,
\fB*\fR for zero or more, \fB+\fR for one or more.  Thick lines
represent strong references; thin lines represent weak references.
.RS -1in
.ps -3
.PS
linethick = 1;
linethick = 1.000000;
box at 2.014690,1.358811 wid 0.317648 height 0.211765 "Bridge"
linethick = 1.000000;
box at 3.264696,0.741178 wid 0.317648 height 0.211765 "Mirror"
linethick = 1.000000;
box at 3.264696,1.994106 wid 0.317648 height 0.211765 "IPFIX"
linethick = 1.000000;
box at 3.264696,0.105883 wid 0.423530 height 0.211765 "Controller"
linethick = 1.000000;
box at 3.264696,2.311754 wid 0.470584 height 0.211765 "AutoAttach"
linethick = 1.000000;
box at 3.264696,1.676459 wid 0.376472 height 0.211765 "NetFlow"
linethick = 1.000000;
box at 3.264696,1.358811 wid 0.317648 height 0.211765 "sFlow"
linethick = 1.000000;
box at 3.264696,0.423530 wid 0.482358 height 0.211765 "Flow_Table"
linethick = 1.000000;
box at 4.400053,0.770570 wid 0.317648 height 0.211765 "Port"
linethick = 0.500000;
box at 5.320384,0.976491 wid 0.317648 height 0.211765 "QoS"
box at 5.320384,0.976491 wid 0.262092 height 0.156209
linethick = 0.500000;
box at 6.341091,0.976491 wid 0.317648 height 0.211765 "Queue"
box at 6.341091,0.976491 wid 0.262092 height 0.156209
linethick = 0.500000;
box at 0.508829,2.047048 wid 1.017658 height 0.211765 "Flow_Sample_Collector_Set"
box at 0.508829,2.047048 wid 0.962102 height 0.156209
linethick = 0.500000;
box at 0.508829,1.041164 wid 0.570580 height 0.211765 "Open_vSwitch"
box at 0.508829,1.041164 wid 0.515024 height 0.156209
linethick = 1.000000;
box at 2.014690,0.723516 wid 0.317648 height 0.211765 "SSL"
linethick = 1.000000;
box at 2.014690,1.041164 wid 0.382354 height 0.211765 "Manager"
linethick = 1.000000;
box at 5.320384,0.658843 wid 0.382354 height 0.211765 "Interface"
linethick = 1.000000;
spline -> from 2.161189,1.252209 to 2.161189,1.252209 to 2.177749,1.236030 to 2.193165,1.218496 to 2.205871,1.199988 to 2.288798,1.079451 to 2.199688,0.982293 to 2.311754,0.888227 to 2.535505,0.700434 to 2.896988,0.700349 to 3.104475,0.719281
"mirrors*" at 2.614705,0.932359
linethick = 1.000000;
spline -> from 2.076483,1.466219 to 2.076483,1.466219 to 2.128874,1.547536 to 2.211377,1.655536 to 2.311754,1.717668 to 2.546898,1.863151 to 2.653543,1.777047 to 2.917656,1.858831 to 2.980042,1.878144 to 3.047214,1.903301 to 3.105703,1.926638
"ipfix?" at 2.614705,1.902963
linethick = 1.000000;
spline -> from 2.166907,1.252675 to 2.166907,1.252675 to 2.182069,1.236581 to 2.195622,1.218962 to 2.205871,1.199988 to 2.303071,1.020665 to 2.172921,0.437676 to 2.311754,0.288238 to 2.498658,0.087090 to 2.833882,0.068163 to 3.052804,0.080958
"controller*" at 2.614705,0.332352
linethick = 1.000000;
spline -> from 2.043193,1.466939 to 2.043193,1.466939 to 2.082793,1.597555 to 2.166568,1.815970 to 2.311754,1.947052 to 2.514243,2.129848 to 2.817999,2.224718 to 3.027858,2.271264
"auto_attach?" at 2.614705,2.285283
linethick = 1.000000;
spline -> from 2.173598,1.423527 to 2.173598,1.423527 to 2.217857,1.440214 to 2.266309,1.457197 to 2.311754,1.470581 to 2.573199,1.547536 to 2.882715,1.609160 to 3.076395,1.644271
"netflow?" at 2.614705,1.649988
linethick = 1.000000;
spline -> from 2.175547,1.358811 to 2.175547,1.358811 to 2.415561,1.358811 to 2.864799,1.358811 to 3.104475,1.358811
"sflow?" at 2.614705,1.402943
linethick = 1.000000;
spline -> from 2.165805,1.252039 to 2.165805,1.252039 to 2.181137,1.236072 to 2.195071,1.218665 to 2.205871,1.199988 to 2.341867,0.965818 to 2.121123,0.792298 to 2.311754,0.600015 to 2.495396,0.414814 to 2.806013,0.389186 to 3.022522,0.397788
"flow_tables value*" at 2.614705,0.644104
linethick = 1.000000;
spline -> from 2.175038,1.317644 to 2.175038,1.317644 to 2.373674,1.267032 to 2.722747,1.179616 to 3.023538,1.111766 to 3.515680,1.000759 to 3.653539,1.032185 to 4.135305,0.882340 to 4.169610,0.871709 to 4.205526,0.858199 to 4.239535,0.844265
"ports*" at 3.264696,1.155898
linethick = 0.500000;
spline -> from 3.424198,0.698443 to 3.424198,0.698443 to 3.482730,0.684763 to 3.549817,0.671464 to 3.611779,0.664688 to 3.843069,0.639530 to 3.906810,0.620726 to 4.135305,0.664688 to 4.170288,0.671464 to 4.206542,0.682773 to 4.240382,0.695563
"select_src_port*" at 3.873521,0.708820
linethick = 0.500000;
spline -> from 3.423859,0.638683 to 3.423859,0.638683 to 3.504626,0.587648 to 3.590179,0.535469 to 3.611779,0.529413 to 3.835784,0.466476 to 3.918330,0.445384 to 4.135305,0.529413 to 4.206034,0.556815 to 4.269606,0.612848 to 4.316618,0.664053
"output_port?" at 3.873521,0.573544
linethick = 0.500000;
spline -> from 3.423647,0.746938 to 3.423647,0.746938 to 3.482730,0.749013 to 3.550283,0.751215 to 3.611779,0.752952 to 3.829304,0.759050 to 4.080712,0.764387 to 4.239959,0.767563
"select_dst_port*" at 3.873521,0.814702
linethick = 1.000000;
spline -> from 4.560571,0.806443 to 4.560571,0.806443 to 4.728712,0.844138 to 4.992148,0.903008 to 5.160713,0.940660
"qos?" at 4.844336,0.950020
linethick = 1.000000;
spline -> from 4.558877,0.731563 to 4.558877,0.731563 to 4.593606,0.724067 to 4.630453,0.716952 to 4.664759,0.711785 to 4.819348,0.688448 to 4.996383,0.674937 to 5.127678,0.667399
"interfaces+" at 4.844336,0.755874
linethick = 1.000000;
spline -> from 5.480055,0.976491 to 5.480055,0.976491 to 5.671067,0.976491 to 5.989561,0.976491 to 6.180997,0.976491
"queues value*" at 5.847255,1.020580
linethick = 1.000000;
spline -> from 0.740839,1.941038 to 0.740839,1.941038 to 1.049380,1.800003 to 1.586882,1.554355 to 1.854214,1.432167
"bridge" at 1.420604,1.802925
linethick = 1.000000;
spline -> from 1.003046,2.153227 to 1.003046,2.153227 to 1.495315,2.240262 to 2.269147,2.325984 to 2.917656,2.170591 to 2.982075,2.155175 to 3.048018,2.125655 to 3.104898,2.095161
"ipfix?" at 2.014690,2.297057
linethick = 1.000000;
spline -> from 0.795601,1.101686 to 0.795601,1.101686 to 1.109649,1.167926 to 1.602045,1.271776 to 1.854172,1.324971
"bridges*" at 1.420604,1.332341
linethick = 1.000000;
spline -> from 0.795601,0.980684 to 0.795601,0.980684 to 1.109649,0.914444 to 1.602045,0.810594 to 1.854172,0.757399
"ssl?" at 1.420604,0.950020
linethick = 1.000000;
spline -> from 0.795601,1.041164 to 0.795601,1.041164 to 1.096858,1.041164 to 1.562190,1.041164 to 1.822195,1.041164
"manager_options*" at 1.420604,1.085296
.ps +3
.PE
.RE\}
.bp
.SH "Open_vSwitch TABLE"
.PP
Configuration for an Open vSwitch daemon\[char46] There must be exactly one record in the \fBOpen_vSwitch\fR table\[char46]
.SS "Summary:
.TQ .25in
\fIConfiguration:\fR
.RS .25in
.TQ 2.75in
\fBbridges\fR
set of \fBBridge\fRs
.TQ 2.75in
\fBssl\fR
optional \fBSSL\fR
.TQ 2.75in
\fBexternal_ids : system-id\fR
optional string
.TQ 2.75in
\fBexternal_ids : xs-system-uuid\fR
optional string
.TQ 2.75in
\fBexternal_ids : hostname\fR
optional string
.TQ 2.75in
\fBexternal_ids : rundir\fR
optional string
.TQ 2.75in
\fBother_config : stats-update-interval\fR
optional string, containing an integer, at least 5,000
.TQ 2.75in
\fBother_config : flow-restore-wait\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : flow-limit\fR
optional string, containing an integer, at least 0
.TQ 2.75in
\fBother_config : max-idle\fR
optional string, containing an integer, at least 500
.TQ 2.75in
\fBother_config : hw-offload\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : tc-policy\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-init\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : dpdk-lcore-mask\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : pmd-cpu-mask\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-alloc-mem\fR
optional string, containing an integer, at least 0
.TQ 2.75in
\fBother_config : dpdk-socket-mem\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-hugepage-dir\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-extra\fR
optional string
.TQ 2.75in
\fBother_config : vhost-sock-dir\fR
optional string
.TQ 2.75in
\fBother_config : vhost-iommu-support\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : tx-flush-interval\fR
optional string, containing an integer, in range 0 to 1,000,000
.TQ 2.75in
\fBother_config : n-handler-threads\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : n-revalidator-threads\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : emc-insert-inv-prob\fR
optional string, containing an integer, in range 0 to 4,294,967,295
.TQ 2.75in
\fBother_config : vlan-limit\fR
optional string, containing an integer, at least 0
.RE
.TQ .25in
\fIStatus:\fR
.RS .25in
.TQ 2.75in
\fBnext_cfg\fR
integer
.TQ 2.75in
\fBcur_cfg\fR
integer
.TQ .25in
\fIStatistics:\fR
.RS .25in
.TQ 2.50in
\fBother_config : enable-statistics\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.50in
\fBstatistics : cpu\fR
optional string, containing an integer, at least 1
.TQ 2.50in
\fBstatistics : load_average\fR
optional string
.TQ 2.50in
\fBstatistics : memory\fR
optional string
.TQ 2.50in
\fBstatistics : process_NAME\fR
optional string
.TQ 2.50in
\fBstatistics : file_systems\fR
optional string
.RE
.RE
.TQ .25in
\fIVersion Reporting:\fR
.RS .25in
.TQ 2.75in
\fBovs_version\fR
optional string
.TQ 2.75in
\fBdb_version\fR
optional string
.TQ 2.75in
\fBsystem_type\fR
optional string
.TQ 2.75in
\fBsystem_version\fR
optional string
.RE
.TQ .25in
\fICapabilities:\fR
.RS .25in
.TQ 2.75in
\fBdatapath_types\fR
set of strings
.TQ 2.75in
\fBiface_types\fR
set of strings
.RE
.TQ .25in
\fIDatabase Configuration:\fR
.RS .25in
.TQ 2.75in
\fBmanager_options\fR
set of \fBManager\fRs
.RE
.TQ .25in
\fICommon Columns:\fR
.RS .25in
.TQ 2.75in
\fBother_config\fR
map of string-string pairs
.TQ 2.75in
\fBexternal_ids\fR
map of string-string pairs
.RE
.SS "Details:
.ST "Configuration:"
.PP
.IP "\fBbridges\fR: set of \fBBridge\fRs"
Set of bridges managed by the daemon\[char46]
.IP "\fBssl\fR: optional \fBSSL\fR"
SSL used globally by the daemon\[char46]
.IP "\fBexternal_ids : system-id\fR: optional string"
A unique identifier for the Open vSwitch\(cqs physical host\[char46] The form of the identifier depends on the type of the host\[char46] On a Citrix XenServer, this will likely be the same as \fBexternal_ids:xs-system-uuid\fR\[char46]
.IP "\fBexternal_ids : xs-system-uuid\fR: optional string"
The Citrix XenServer universally unique identifier for the physical host as displayed by \fBxe host\-list\fR\[char46]
.IP "\fBexternal_ids : hostname\fR: optional string"
The hostname for the host running Open vSwitch\[char46] This is a fully qualified domain name since version 2\[char46]6\[char46]2\[char46]
.IP "\fBexternal_ids : rundir\fR: optional string"
In Open vSwitch 2\[char46]8 and later, the run directory of the running Open vSwitch daemon\[char46] This directory is used for runtime state such as control and management sockets\[char46] The value of \fBother_config:vhost-sock-dir\fR is relative to this directory\[char46]
.IP "\fBother_config : stats-update-interval\fR: optional string, containing an integer, at least 5,000"
Interval for updating statistics to the database, in milliseconds\[char46] This option will affect the update of the \fBstatistics\fR column in the following tables: \fBPort\fR, \fBInterface
\fR, \fBMirror\fR\[char46]
.IP
Default value is 5000 ms\[char46]
.IP
Getting statistics more frequently can be achieved via OpenFlow\[char46]
.IP "\fBother_config : flow-restore-wait\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
When \fBovs\-vswitchd\fR starts up, it has an empty flow table and therefore it handles all arriving packets in its default fashion according to its configuration, by dropping them or sending them to an OpenFlow controller or switching them as a standalone switch\[char46] This behavior is ordinarily desirable\[char46] However, if \fBovs\-vswitchd\fR is restarting as part of a ``hot-upgrade,\(cq\(cq then this leads to a relatively long period during which packets are mishandled\[char46]
.IP
This option allows for improvement\[char46] When \fBovs\-vswitchd\fR starts with this value set as \fBtrue\fR, it will neither flush or expire previously set datapath flows nor will it send and receive any packets to or from the datapath\[char46] When this value is later set to \fBfalse\fR, \fBovs\-vswitchd\fR will start receiving packets from the datapath and re-setup the flows\[char46]
.IP
Thus, with this option, the procedure for a hot-upgrade of \fBovs\-vswitchd\fR becomes roughly the following:
.RS
.IP 1. .25in
Stop \fBovs\-vswitchd\fR\[char46]
.IP 2. .25in
Set \fBother_config:flow-restore-wait\fR to \fBtrue\fR\[char46]
.IP 3. .25in
Start \fBovs\-vswitchd\fR\[char46]
.IP 4. .25in
Use \fBovs\-ofctl\fR (or some other program, such as an OpenFlow controller) to restore the OpenFlow flow table to the desired state\[char46]
.IP 5. .25in
Set \fBother_config:flow-restore-wait\fR to \fBfalse\fR (or remove it entirely from the database)\[char46]
.RE
.IP
The \fBovs\-ctl\fR\(cqs ``restart\(cq\(cq and ``force-reload-kmod\(cq\(cq functions use the above config option during hot upgrades\[char46]
.IP "\fBother_config : flow-limit\fR: optional string, containing an integer, at least 0"
The maximum number of flows allowed in the datapath flow table\[char46] Internally OVS will choose a flow limit which will likely be lower than this number, based on real time network conditions\[char46] Tweaking this value is discouraged unless you know exactly what you\(cqre doing\[char46]
.IP
The default is 200000\[char46]
.IP "\fBother_config : max-idle\fR: optional string, containing an integer, at least 500"
The maximum time (in ms) that idle flows will remain cached in the datapath\[char46] Internally OVS will check the validity and activity for datapath flows regularly and may expire flows quicker than this number, based on real time network conditions\[char46] Tweaking this value is discouraged unless you know exactly what you\(cqre doing\[char46]
.IP
The default is 10000\[char46]
.IP "\fBother_config : hw-offload\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
Set this value to \fBtrue\fR to enable netdev flow offload\[char46]
.IP
The default value is \fBfalse\fR\[char46] Changing this value requires restarting the daemon
.IP
Currently Open vSwitch supports hardware offloading on Linux systems\[char46] On other systems, this value is ignored\[char46] This functionality is considered \(cqexperimental\(cq\[char46] Depending on which OpenFlow matches and actions are configured, which kernel version is used, and what hardware is available, Open vSwitch may not be able to offload functionality to hardware\[char46]
.IP "\fBother_config : tc-policy\fR: optional string"
Specified the policy used with HW offloading\[char46] Options: \fBnone\fR - Add software rule and offload rule to HW\[char46] \fBskip_sw\fR - Offload rule to HW only\[char46] \fBskip_hw\fR - Add software rule without offloading rule to HW\[char46]
.IP
This is only relevant if HW offloading is enabled (hw-offload)\[char46]
.IP
The default value is \fBnone\fR\[char46]
.IP "\fBother_config : dpdk-init\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
Set this value to \fBtrue\fR to enable runtime support for DPDK ports\[char46] The vswitch must have compile-time support for DPDK as well\[char46]
.IP
The default value is \fBfalse\fR\[char46] Changing this value requires restarting the daemon
.IP
If this value is \fBfalse\fR at startup, any dpdk ports which are configured in the bridge will fail due to memory errors\[char46]
.IP "\fBother_config : dpdk-lcore-mask\fR: optional string, containing an integer, at least 1"
Specifies the CPU cores where dpdk lcore threads should be spawned\[char46] The DPDK lcore threads are used for DPDK library tasks, such as library internal message processing, logging, etc\[char46] Value should be in the form of a hex string (so \(cq0x123\(cq) similar to the \(cqtaskset\(cq mask input\[char46]
.IP
The lowest order bit corresponds to the first CPU core\[char46] A set bit means the corresponding core is available and an lcore thread will be created and pinned to it\[char46] If the input does not cover all cores, those uncovered cores are considered not set\[char46]
.IP
For performance reasons, it is best to set this to a single core on the system, rather than allow lcore threads to float\[char46]
.IP
If not specified, the value will be determined by choosing the lowest CPU core from initial cpu affinity list\[char46] Otherwise, the value will be passed directly to the DPDK library\[char46]
.IP "\fBother_config : pmd-cpu-mask\fR: optional string"
Specifies CPU mask for setting the cpu affinity of PMD (Poll Mode Driver) threads\[char46] Value should be in the form of hex string, similar to the dpdk EAL \(cq-c COREMASK\(cq option input or the \(cqtaskset\(cq mask input\[char46]
.IP
The lowest order bit corresponds to the first CPU core\[char46] A set bit means the corresponding core is available and a pmd thread will be created and pinned to it\[char46] If the input does not cover all cores, those uncovered cores are considered not set\[char46]
.IP
If not specified, one pmd thread will be created for each numa node and pinned to any available core on the numa node by default\[char46]
.IP "\fBother_config : dpdk-alloc-mem\fR: optional string, containing an integer, at least 0"
Specifies the amount of memory to preallocate from the hugepage pool, regardless of socket\[char46] It is recommended that dpdk-socket-mem is used instead\[char46]
.IP "\fBother_config : dpdk-socket-mem\fR: optional string"
Specifies the amount of memory to preallocate from the hugepage pool, on a per-socket basis\[char46]
.IP
The specifier is a comma-separated string, in ascending order of CPU socket\[char46] E\[char46]g\[char46] On a four socket system 1024,0,2048 would set socket 0 to preallocate 1024MB, socket 1 to preallocate 0MB, socket 2 to preallocate 2048MB and socket 3 (no value given) to preallocate 0MB\[char46]
.IP
If dpdk-socket-mem and dpdk-alloc-mem are not specified, dpdk-socket-mem will be used and the default value is 1024,0\[char46] If dpdk-socket-mem and dpdk-alloc-mem are specified at same time, dpdk-socket-mem will be used as default\[char46] Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : dpdk-hugepage-dir\fR: optional string"
Specifies the path to the hugetlbfs mount point\[char46]
.IP
If not specified, this will be guessed by the DPDK library (default is /dev/hugepages)\[char46] Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : dpdk-extra\fR: optional string"
Specifies additional eal command line arguments for DPDK\[char46]
.IP
The default is empty\[char46] Changing this value requires restarting the daemon
.IP "\fBother_config : vhost-sock-dir\fR: optional string"
Specifies a relative path from \fBexternal_ids:rundir\fR to the vhost-user unix domain socket files\[char46] If this value is unset, the sockets are put directly in \fBexternal_ids:rundir\fR\[char46]
.IP
Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : vhost-iommu-support\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
vHost IOMMU is a security feature, which restricts the vhost memory that a virtio device may access\[char46] vHost IOMMU support is disabled by default, due to a bug in QEMU implementations of the vhost REPLY_ACK protocol, (on which vHost IOMMU relies) prior to v2\[char46]9\[char46]1\[char46] Setting this value to \fBtrue\fR enables vHost IOMMU support for vHost User Client ports in OvS-DPDK, starting from DPDK v17\[char46]11\[char46]
.IP
Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : tx-flush-interval\fR: optional string, containing an integer, in range 0 to 1,000,000"
Specifies the time in microseconds that a packet can wait in output batch for sending i\[char46]e\[char46] amount of time that packet can spend in an intermediate output queue before sending to netdev\[char46] This option can be used to configure balance between throughput and latency\[char46] Lower values decreases latency while higher values may be useful to achieve higher performance\[char46]
.IP
Defaults to 0 i\[char46]e\[char46] instant packet sending (latency optimized)\[char46]
.IP "\fBother_config : n-handler-threads\fR: optional string, containing an integer, at least 1"
Specifies the number of threads for software datapaths to use for handling new flows\[char46] The default the number of online CPU cores minus the number of revalidators\[char46]
.IP
This configuration is per datapath\[char46] If you have more than one software datapath (e\[char46]g\[char46] some \fBsystem\fR bridges and some \fBnetdev\fR bridges), then the total number of threads is \fBn\-handler\-threads\fR times the number of software datapaths\[char46]
.IP "\fBother_config : n-revalidator-threads\fR: optional string, containing an integer, at least 1"
Specifies the number of threads for software datapaths to use for revalidating flows in the datapath\[char46] Typically, there is a direct correlation between the number of revalidator threads, and the number of flows allowed in the datapath\[char46] The default is the number of cpu cores divided by four plus one\[char46] If \fBn\-handler\-threads\fR is set, the default changes to the number of cpu cores minus the number of handler threads\[char46]
.IP

LUONGTD/Open-vSwitch/code/vswitchd/ovs-vswitchd.c
/* Copyright (c) 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016 Nicira, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#ifdef HAVE_MLOCKALL
#include <sys/mman.h>
#endif

#include "bridge.h"
#include "command-line.h"
#include "compiler.h"
#include "daemon.h"
#include "dirs.h"
#include "dpif.h"
#include "dummy.h"
#include "fatal-signal.h"
#include "memory.h"
#include "netdev.h"
#include "openflow/openflow.h"
#include "ovsdb-idl.h"
#include "ovs-rcu.h"
#include "ovs-router.h"
#include "ovs-thread.h"
#include "openvswitch/poll-loop.h"
#include "simap.h"
#include "stream-ssl.h"
#include "stream.h"
#include "svec.h"
#include "timeval.h"
#include "unixctl.h"
#include "util.h"
#include "openvswitch/vconn.h"
#include "openvswitch/vlog.h"
#include "lib/vswitch-idl.h"
#include "lib/dns-resolve.h"

VLOG_DEFINE_THIS_MODULE(vswitchd);

/* --mlockall: If set, locks all process memory into physical RAM, preventing
 * the kernel from paging any of its memory to disk. */
static bool want_mlockall;

static unixctl_cb_func ovs_vswitchd_exit;

static char *parse_options(int argc, char *argv[], char **unixctl_path);
OVS_NO_RETURN static void usage(void);

struct ovs_vswitchd_exit_args {
    bool *exiting;
    bool *cleanup;
};

int main(int argc, char *argv[]) {
    ...
    /* Step.1: init bridge module, obtain config from ovsdb*/
    bridge_init(remote);

    /* Step.2: daemon loop*/
    while (!exiting) {
        ...
        /* Step.2.1: process control messages from OpenFlow Controller and CLIENT*/
        bridge_run();
            | 
            | -- dpdk_init()
            | -- bridge_init_ofproto()          // init bridge, only once
            | -- bridge_run__()
                |
                | --for (datapath types): /* Let each datapath type do the work that it need to do. */
                |       ofproto_type_run(type)
                | --for (all_bridges):
                |       ofproto_run(bridge) /* Handle messages from OpenFlow Controller*/

        unixctl_server_run(unixctl); /* Recieve control messages from CLI (ovs-appctl1 ...) */
        netdev_run(); /* Performs periodic work needed by all the various kinds of netdevs */

        /* Step2.2: Wait events arriving*/
        memory_wait();
        bridge_wait();
        unixctl_server_wait(unixctl);
        netdev_wait();
        ...
        /* Step.2.3: block util events arrive*/
        poll_block();
        ...
    }
    return 0;
}

static char *
parse_options(int argc, char *argv[], char **unixctl_pathp)
{
    enum {
        OPT_PEER_CA_CERT = UCHAR_MAX + 1,
        OPT_MLOCKALL,
        OPT_UNIXCTL,
        VLOG_OPTION_ENUMS,
        OPT_BOOTSTRAP_CA_CERT,
        OPT_ENABLE_DUMMY,
        OPT_DISABLE_SYSTEM,
        DAEMON_OPTION_ENUMS,
        OPT_DPDK,
        SSL_OPTION_ENUMS,
        OPT_DUMMY_NUMA,
    };
    static const struct option long_options[] = {
        {"help",        no_argument, NULL, 'h'},
        {"version",     no_argument, NULL, 'V'},
        {"mlockall",    no_argument, NULL, OPT_MLOCKALL},
        {"unixctl",     required_argument, NULL, OPT_UNIXCTL},
        DAEMON_LONG_OPTIONS,
        VLOG_LONG_OPTIONS,
        STREAM_SSL_LONG_OPTIONS,
        {"peer-ca-cert", required_argument, NULL, OPT_PEER_CA_CERT},
        {"bootstrap-ca-cert", required_argument, NULL, OPT_BOOTSTRAP_CA_CERT},
        {"enable-dummy", optional_argument, NULL, OPT_ENABLE_DUMMY},
        {"disable-system", no_argument, NULL, OPT_DISABLE_SYSTEM},
        {"dpdk", optional_argument, NULL, OPT_DPDK},
        {"dummy-numa", required_argument, NULL, OPT_DUMMY_NUMA},
        {NULL, 0, NULL, 0},
    };
    char *short_options = ovs_cmdl_long_options_to_short_options(long_options);

    for (;;) {
        int c;

        c = getopt_long(argc, argv, short_options, long_options, NULL);
        if (c == -1) {
            break;
        }

        switch (c) {
        case 'h':
            usage();

        case 'V':
            ovs_print_version(0, 0);
            print_dpdk_version();
            exit(EXIT_SUCCESS);

        case OPT_MLOCKALL:
            want_mlockall = true;
            break;

        case OPT_UNIXCTL:
            *unixctl_pathp = optarg;
            break;

        VLOG_OPTION_HANDLERS
        DAEMON_OPTION_HANDLERS
        STREAM_SSL_OPTION_HANDLERS

        case OPT_PEER_CA_CERT:
            stream_ssl_set_peer_ca_cert_file(optarg);
            break;

        case OPT_BOOTSTRAP_CA_CERT:
            stream_ssl_set_ca_cert_file(optarg, true);
            break;

        case OPT_ENABLE_DUMMY:
            dummy_enable(optarg);
            break;

        case OPT_DISABLE_SYSTEM:
            dp_blacklist_provider("system");
            ovs_router_disable_system_routing_table();
            break;

        case '?':
            exit(EXIT_FAILURE);

        case OPT_DPDK:
            ovs_fatal(0, "Using --dpdk to configure DPDK is not supported.");
            break;

        case OPT_DUMMY_NUMA:
            ovs_numa_set_dummy(optarg);
            break;

        default:
            abort();
        }
    }
    free(short_options);

    argc -= optind;
    argv += optind;

    switch (argc) {
    case 0:
        return xasprintf("unix:%s/db.sock", ovs_rundir());

    case 1:
        return xstrdup(argv[0]);

    default:
        VLOG_FATAL("at most one non-option argument accepted; "
                   "use --help for usage");
    }
}

static void usage(void)
{
    printf("%s: Open vSwitch daemon\n"
           "usage: %s [OPTIONS] [DATABASE]\n"
           "where DATABASE is a socket on which ovsdb-server is listening\n"
           "      (default: \"unix:%s/db.sock\").\n",
           program_name, program_name, ovs_rundir());
    stream_usage("DATABASE", true, false, true);
    daemon_usage();
    vlog_usage();
    printf("\nDPDK options:\n"
           "Configuration of DPDK via command-line is removed from this\n"
           "version of Open vSwitch. DPDK is configured through ovsdb.\n"
          );
    printf("\nOther options:\n"
           "  --unixctl=SOCKET          override default control socket name\n"
           "  -h, --help                display this help message\n"
           "  -V, --version             display version information\n");
    exit(EXIT_SUCCESS);
}

static void
ovs_vswitchd_exit(struct unixctl_conn *conn, int argc,
                  const char *argv[], void *exit_args_)
{
    struct ovs_vswitchd_exit_args *exit_args = exit_args_;
    *exit_args->exiting = true;
    *exit_args->cleanup = argc == 2 && !strcmp(argv[1], "--cleanup");
    unixctl_command_reply(conn, NULL);
}
